CWE-330: Use of Insufficiently Random Values Documentation

BartyBoi1128 opened the pull request:
Documentation for CWE-330, and a few minor code changes

myteron commented on docs/Secure-Coding-Guide-for-Python/CWE-693/CWE-330/README.md:
wrong ref, their publication is public available.

```suggestion
[MATSUMOTO, NISHIMURA 1998] Mersenne Twister: A 623-Dimensionally
Equidistributed Uniform Pseudo-Random
Number Generator, available from: [https://dl.acm.org/doi/pdf/10.1145/272991.272995](https://dl.acm.org/doi/pdf/10.1145/272991.272995), [accessed December 2024]|
```

myteron commented on docs/Secure-Coding-Guide-for-Python/CWE-693/CWE-330/README.md:
```diff
@@ -0,0 +1,97 @@
+
+PRNGs suitable for encryption must mix non-computational sources such as a mouse, keyboard, or even  Lava Lamps [LavaRnd] to be random enough for encryption.
+
+Python's random module is a standard library module that provides functions to generate pseudorandom numbers for various distributions. This module however can lead to a vulnerability due to its predictability. The random module is based on the Mersenne Twister [Mersenne Twister - an overview | ScienceDirect Topics], which is a deterministic algorithm, hence, if an attacker knows or can guess the ...
```
however is a word reserved for AIs ;P, updated link to match references linking directly to their publication assuming that is what we want.

```suggestion
Python's random module is a standard library module that provides functions to generate pseudorandom numbers for various distributions. This module can lead to a vulnerability due to its predictability. The random module is based on the Mersenne Twister `MT19937`
[MATSUMOTO, NISHIMURA 1998] (https://dl.acm.org/doi/pdf/10.1145/272991.272995), which is a deterministic algorithm, hence, if an attacker knows or can guess the seed value, they can predict the entire sequence of the pseudorandom numbers. This also means that if two Random class objects are created using an identical seed, they will generate the same sequence of numbers, regardless of the Python environment.
```

myteron commented on docs/Secure-Coding-Guide-for-Python/CWE-693/CWE-330/README.md:
```diff
@@ -0,0 +1,97 @@
+
+Therefore, the random module is unsuitable for applications requiring high security as it does not incorporate cryptographic randomness, which means it is not resistant to reverse engineering. Its limited entropy makes it easier for attackers to deduce the internal state of the generator and predict future outputs.
+
+Instead, for generating random numbers, it is recommended to use a more robust option, like Python's secrets module.
```
```suggestion
Instead, for generating random numbers, it is recommended to use a more robust option, such as Python's `secrets` module.

```

myteron commented on docs/Secure-Coding-Guide-for-Python/CWE-693/CWE-330/README.md:
```suggestion
 The `compliant01.py` solution uses the `secrets` module to generate the random numbers. The `secrets` module provides access to the most secure source of randomness that an OS provides through `os.urandom()`.

```

myteron commented on docs/Secure-Coding-Guide-for-Python/CWE-693/CWE-330/README.md:
FLP02-CPP is not accessible, suggest to remove it.

```suggestion
```

myteron reviewed (changes requested):
Loads of empty lines with a single space. Please install black, format code and C&P into readme.md.

Suggest to update the MT to the publication, that is assuming the existing reference did not provide any more insights. 

some formatting issues

myteron commented on docs/Secure-Coding-Guide-for-Python/CWE-693/CWE-330/README.md:
```diff
@@ -0,0 +1,101 @@
+PRNGs suitable for encryption must mix non-computational sources such as a mouse, keyboard, or even  Lava Lamps [LavaRnd] to be random enough for encryption.
+
+Python's random module is a standard library module that provides functions to generate pseudorandom numbers for various distributions. This module can lead to a vulnerability due to its predictability. The random module is based on the Mersenne Twister `MT19937`
+[MATSUMOTO, NISHIMURA 1998] (https://dl.acm.org/doi/pdf/10.1145/272991.272995), which is a deterministic algorithm, that, given a particular input, will always produce the same output [Wikipedia 2024](https://en.wikipedia.org/wiki/Deterministic_algorithm). An attacker knowing or can guessing the seed value can predict the entire sequence of the pseudorandom numbers. This also means that if two Ra ...
```
I messed up the formatting

```suggestion
[[MATSUMOTO, NISHIMURA 1998](https://dl.acm.org/doi/pdf/10.1145/272991.272995)], which is a deterministic algorithm, that, given a particular input, will always produce the same output [[Wikipedia 2024](https://en.wikipedia.org/wiki/Deterministic_algorithm)]. An attacker knowing or can guessing the seed value can predict the entire sequence of the pseudorandom numbers. This also means that if two Random class objects are created using an identical seed, they will generate the same sequence of numbers, regardless of the Python environment.
```

myteron commented on docs/Secure-Coding-Guide-for-Python/CWE-693/CWE-330/README.md:
broken formatting

```suggestion
|[MATSUMOTO, NISHIMURA 1998]|Mersenne Twister: A 623-Dimensionally Equidistributed Uniform Pseudo-Random Number Generator, available from: [https://dl.acm.org/doi/pdf/10.1145/272991.272995](https://dl.acm.org/doi/pdf/10.1145/272991.272995), [accessed December 2024]|
```

myteron commented on docs/Secure-Coding-Guide-for-Python/CWE-693/CWE-330/noncompliant01.py:
```diff
@@ -2,13 +2,13 @@
-
+ 
+ 
 def generate_web_token():
     """Poor random number generator"""
     return random.randrange(int("1" + "0" * 31), int("9" * 32), 1)
-
-
+ 
+ 
 #####################
```
spaces removed
```suggestion


def generate_web_token():
    """Poor random number generator"""
    return random.randrange(int("1" + "0" * 31), int("9" * 32), 1)


#####################
```

myteron commented on docs/Secure-Coding-Guide-for-Python/CWE-693/CWE-330/README.md:
```diff
@@ -0,0 +1,101 @@
+
+Instead, for generating random numbers, it is recommended to use a more robust option, such as Python's `secrets` module.
+
+
```
double blank line
```suggestion
```

myteron commented on docs/Secure-Coding-Guide-for-Python/CWE-693/CWE-330/noncompliant01.py:
```diff
@@ -2,13 +2,13 @@
-
+ 
+ 
 def generate_web_token():
     """Poor random number generator"""
     return random.randrange(int("1" + "0" * 31), int("9" * 32), 1)
-
-
+ 
+ 
```
remove empty space
```suggestion


def generate_web_token():
    """Poor random number generator"""
    return random.randrange(int("1" + "0" * 31), int("9" * 32), 1)


```

myteron commented on docs/Secure-Coding-Guide-for-Python/CWE-693/CWE-330/README.md:
remove empty space
```suggestion


def generate_web_token():
    """Poor random number generator"""
    return random.randrange(int("1" + "0" * 31), int("9" * 32), 1)


```

myteron commented on docs/Secure-Coding-Guide-for-Python/CWE-693/CWE-330/README.md:
remove double blank
```suggestion
```

myteron commented on docs/Secure-Coding-Guide-for-Python/CWE-693/CWE-330/README.md:
empty space removed
```suggestion


def generate_web_token():
    """Better cryptographic number generator"""
    return secrets.token_urlsafe()


```

myteron commented on docs/Secure-Coding-Guide-for-Python/CWE-693/CWE-330/compliant01.py:
for some reason I can't make a suggestion, you need to remove the empty spaces from 4 blank lines

myteron replied:
done this now - resolved

myteron reviewed (changes requested):
some formatting issues

myteron reviewed (approved):
All formatting fixed. no major content changes as such, only added sonar ref back in.

s19110 commented on docs/Secure-Coding-Guide-for-Python/CWE-693/CWE-330/README.md:
```diff
@@ -0,0 +1,99 @@
+PRNGs suitable for encryption must mix non-computational sources such as a mouse, keyboard, or even  Lava Lamps [LavaRnd] to be random enough for encryption.
+
+Python's random module is a standard library module that provides functions to generate pseudorandom numbers for various distributions. This module can lead to a vulnerability due to its predictability. The random module is based on the Mersenne Twister `MT19937`
+[[MATSUMOTO, NISHIMURA 1998](https://dl.acm.org/doi/pdf/10.1145/272991.272995)], which is a deterministic algorithm, that, given a particular input, will always produce the same output [[Wikipedia 2024](https://en.wikipedia.org/wiki/Deterministic_algorithm)]. An attacker knowing or can guessing the seed value can predict the entire sequence of the pseudorandom numbers. This also means that if two ...
```
Grammatical error + I think adding backticks to `Random` would make it clearer that this is the name of the class:
```suggestion
[[MATSUMOTO, NISHIMURA 1998](https://dl.acm.org/doi/pdf/10.1145/272991.272995)], which is a deterministic algorithm, that, given a particular input, will always produce the same output [[Wikipedia 2024](https://en.wikipedia.org/wiki/Deterministic_algorithm)]. An attacker knowing or guessing the seed value can predict the entire sequence of the pseudorandom numbers. This also means that if two `Random` class objects are created using an identical seed, they will generate the same sequence of numbers, regardless of the Python environment.

```

s19110 commented on docs/Secure-Coding-Guide-for-Python/CWE-693/CWE-330/README.md:
```diff
@@ -0,0 +1,99 @@
+
+PRNGs suitable for encryption must mix non-computational sources such as a mouse, keyboard, or even  Lava Lamps [LavaRnd] to be random enough for encryption.
+
+Python's random module is a standard library module that provides functions to generate pseudorandom numbers for various distributions. This module can lead to a vulnerability due to its predictability. The random module is based on the Mersenne Twister `MT19937`
```
Backticks around `random` would make it clearer that it's the name of the module and not some unspecified random module.
```suggestion
Python's `random` module is a standard library module that provides functions to generate pseudorandom numbers for various distributions. This module can lead to a vulnerability due to its predictability. The random module is based on the Mersenne Twister `MT19937`
```

s19110 commented on docs/Secure-Coding-Guide-for-Python/CWE-693/CWE-330/README.md:
```diff
@@ -0,0 +1,99 @@
+Python's random module is a standard library module that provides functions to generate pseudorandom numbers for various distributions. This module can lead to a vulnerability due to its predictability. The random module is based on the Mersenne Twister `MT19937`
+[[MATSUMOTO, NISHIMURA 1998](https://dl.acm.org/doi/pdf/10.1145/272991.272995)], which is a deterministic algorithm, that, given a particular input, will always produce the same output [[Wikipedia 2024](https://en.wikipedia.org/wiki/Deterministic_algorithm)]. An attacker knowing or can guessing the seed value can predict the entire sequence of the pseudorandom numbers. This also means that if two ...
+
+Therefore, the random module is unsuitable for applications requiring high security as it does not incorporate cryptographic randomness, which means it is not resistant to reverse engineering. Its limited entropy makes it easier for attackers to deduce the internal state of the generator and predict future outputs.
```
Same case as the line 9 comment.
```suggestion
Therefore, the `random` module is unsuitable for applications requiring high security as it does not incorporate cryptographic randomness, which means it is not resistant to reverse engineering. Its limited entropy makes it easier for attackers to deduce the internal state of the generator and predict future outputs.
```

s19110 commented on docs/Secure-Coding-Guide-for-Python/CWE-693/CWE-330/README.md:
The Wikipedia reference was missing in the bibliography.
```suggestion
|[MATSUMOTO, NISHIMURA 1998]|Mersenne Twister: A 623-Dimensionally Equidistributed Uniform Pseudo-Random Number Generator [online]. Available from: [https://dl.acm.org/doi/pdf/10.1145/272991.272995](https://dl.acm.org/doi/pdf/10.1145/272991.272995) [accessed 12 December 2024]|
|[[Wikipedia 2024](https://en.wikipedia.org/wiki/Deterministic_algorithm)]|Deterministic algorithm [online]. Available from: [https://en.wikipedia.org/wiki/Deterministic_algorithm](https://en.wikipedia.org/wiki/Deterministic_algorithm) [accessed 12 December 2024]|
```

david-a-wheeler commented on docs/Secure-Coding-Guide-for-Python/CWE-693/CWE-330/README.md:
```diff
@@ -0,0 +1,100 @@
+Python's `random` module is a standard library module that provides functions to generate pseudorandom numbers for various distributions. This module can lead to a vulnerability due to its predictability. The random module is based on the Mersenne Twister `MT19937`
+[[MATSUMOTO, NISHIMURA 1998](https://dl.acm.org/doi/pdf/10.1145/272991.272995)], which is a deterministic algorithm, that, given a particular input, will always produce the same output [[Wikipedia 2024](https://en.wikipedia.org/wiki/Deterministic_algorithm)]. An attacker knowing or guessing the seed value can predict the entire sequence of the pseudorandom numbers. This also means that if two `Ra ...
+
+Therefore, the `random` module is unsuitable for applications requiring high security as it does not incorporate cryptographic randomness, which means it is not resistant to reverse engineering. Its limited entropy makes it easier for attackers to deduce the internal state of the generator and predict future outputs.
```
```suggestion
Therefore, the `random` module is unsuitable for applications requiring security as it does not incorporate cryptographic randomness, which means it is predictable. Its use makes it easy for attackers to deduce the internal state of the generator and predict future outputs.

```

"Reverse engineering" isn't the issue. The problem is that you're using the wrong algorithm.

david-a-wheeler commented on docs/Secure-Coding-Guide-for-Python/CWE-693/CWE-330/README.md:
```diff
@@ -0,0 +1,100 @@
+
+Therefore, the `random` module is unsuitable for applications requiring high security as it does not incorporate cryptographic randomness, which means it is not resistant to reverse engineering. Its limited entropy makes it easier for attackers to deduce the internal state of the generator and predict future outputs.
+
+Instead, for generating random numbers, it is recommended to use a more robust option, such as Python's `secrets` module.
```
```suggestion
Instead, for generating random numbers for security purposes, use an appropriate option, such as Python's `secrets` module.

```

The `random` algorithm is robust, it's just robustly wrong when you use it for the wrong purpose.

BartyBoi1128 commented:
Hi @david-a-wheeler 

I have addressed your comments, could you have a look and see if you are happy with the current wording?

Kind regards,
Bart.
